tools/libSwiftScan/SwiftCaching.cpp

```diff
@@ -0,0 +1,838 @@
+}
+
+swiftscan_cache_replay_instance_t
+swiftscan_cache_replay_instance_create(int argc, const char **argv,
+                                       swiftscan_string_ref_t *error) {
+  auto *Instance = new SwiftScanReplayInstance();
+  llvm::StringSaver Saver(Instance->StringAlloc);
+  for (int i = 0; i < argc; ++i) {
+    auto Str = Saver.save(argv[i]);
+    Instance->Args.push_back(Str.data());
+  }
+
+  // Capture the diagnostics when creating invocation.
+  std::string err_msg;
+  llvm::raw_string_ostream OS(err_msg);
+  swift::PrintingDiagnosticConsumer Diags(OS);
+  swift::SourceManager SrcMgr;
+  swift::DiagnosticEngine DE(SrcMgr);
+  DE.addConsumer(Diags);
+
+  std::string MainExecutablePath = llvm::sys::fs::getMainExecutable(
+      "swift-frontend", (void *)swiftscan_cache_replay_compilation);
+
+  if (Instance->Invocation.parseArgs(Instance->Args, DE, nullptr, {},
+                                     MainExecutablePath)) {
+    delete Instance;
+    *error = swift::c_string_utils::create_clone(err_msg.c_str());
+    return nullptr;
+  }
+
+  if (!Instance->Invocation.getFrontendOptions().EnableCaching) {
+    delete Instance;
+    *error = swift::c_string_utils::create_clone(
+        "caching is not enabled from command-line");
+    return nullptr;
+  }
+
+  return wrap(Instance);
+}
+
+void swiftscan_cache_replay_instance_dispose(
+    swiftscan_cache_replay_instance_t instance) {
+  delete unwrap(instance);
+}
+
+namespace {
+class StreamOutputFileImpl final
+    : public llvm::RTTIExtends<StreamOutputFileImpl,
+                               llvm::vfs::OutputFileImpl> {
+
+public:
+  StreamOutputFileImpl(llvm::raw_pwrite_stream &OS) : OS(OS) {}
+  llvm::Error keep() final { return llvm::Error::success(); }
+  llvm::Error discard() final { return llvm::Error::success(); }
+  llvm::raw_pwrite_stream &getOS() final { return OS; }
+
+private:
+  llvm::raw_pwrite_stream &OS;
+};
+
+// The replay output backend. Currently, it redirects "-" to the
+// stdout provided.
+struct ReplayOutputBackend : public llvm::vfs::ProxyOutputBackend {
+  llvm::Expected<std::unique_ptr<llvm::vfs::OutputFileImpl>>
+  createFileImpl(llvm::StringRef Path,
+                 std::optional<llvm::vfs::OutputConfig> Config) override {
+    if (Path == "-")
+      return std::make_unique<StreamOutputFileImpl>(StdOut);
+    return llvm::vfs::ProxyOutputBackend::createFileImpl(Path, Config);
+  }
+
+  llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend>
+  cloneImpl() const override {
+    return llvm::makeIntrusiveRefCnt<ReplayOutputBackend>(
+        getUnderlyingBackend().clone(), StdOut);
+  }
+
+  ReplayOutputBackend(
+      llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> UnderlyingBackend,
+      llvm::raw_pwrite_stream &StdOut)
+      : ProxyOutputBackend(UnderlyingBackend), StdOut(StdOut) {}
+
+  llvm::raw_pwrite_stream &StdOut;
+};
+
+llvm::Error replayCompilation(SwiftScanReplayInstance &Instance,
```

Comment:
Even though this is in an anonymous namespace, llvm style recommends to still use `static` for top-level functions for clarity.